### PR TITLE
Concurrent tables creation in `recoverLostReplica`

### PR DESCRIPTION
**Pull request by thevar1able**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Allow concurrent table creation in `DatabaseReplicated` during `recoverLostReplica`.

### CODE REVIEW

File: src/Common/CurrentMetrics.cpp [hunk 1]
```diff
@@ -176,6 +176,9 @@
     M(MergeTreePartsCleanerThreads, "Number of threads in the MergeTree parts cleaner thread pool.") \
     M(MergeTreePartsCleanerThreadsActive, "Number of threads in the MergeTree parts cleaner thread pool running a task.") \
     M(MergeTreePartsCleanerThreadsScheduled, "Number of queued or active jobs in the MergeTree parts cleaner thread pool.") \
+    M(CreateTablesThreads, "Number of threads in the threadpool for loading Outdated data parts.") \
```
Comment (thevar1able) [inline src/Common/CurrentMetrics.cpp:179]: Don't forget to change descriptions

**Comment by robot-clickhouse-ci-1** [comment]
*This is an automated comment for commit 4715e2276e15b2a4bb94047c5fd12d92a8e34ad3 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/59277/4715e2276e15b2a4bb94047c5fd12d92a8e34ad3/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/59277/4715e2276e15b2a4bb94047c5fd12d92a8e34ad3/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

**Review by alesapin — APPROVED** [review]
Good, let's enable in CI by default before merge?

File: src/Databases/TablesDependencyGraph.cpp [hunk 2]
```diff
@@ -699,6 +699,17 @@ std::vector<StorageID> TablesDependencyGraph::getTablesSortedByDependency() cons
 }
 
 
+std::map<size_t, std::vector<StorageID>> TablesDependencyGraph::getTablesSortedByDependencyWithLevels() const
```
Comment (tavplubix) [inline src/Databases/TablesDependencyGraph.cpp:702]: Why not `std::vector<std::vector<StorageID>>`?

File: src/Core/ServerSettings.h [hunk 3]
```diff
@@ -114,6 +114,7 @@ namespace DB
     M(Bool, validate_tcp_client_information, false, "Validate client_information in the query packet over the native TCP protocol.", 0) \
     M(Bool, storage_metadata_write_full_object_key, false, "Write disk metadata files with VERSION_FULL_OBJECT_KEY format", 0) \
     M(UInt64, max_materialized_views_count_for_table, 0, "A limit on the number of materialized views attached to a table.", 0) \
+    M(UInt64, max_database_replicated_create_table_thread_pool_size, 0, "The number of threads to create tables during replica recovery in DatabaseReplicated. Zero means tables will be created sequentially.", 0) \
```
Comment (tavplubix) [inline src/Core/ServerSettings.h:117]: > Zero means tables will be created sequentially

One means the same. Usually, we use 0 as a special value that means "choose the number of threads automatically (according to the number of cores)"

File: tests/config/config.d/database_replicated.xml [hunk 4]
```diff
@@ -97,4 +97,5 @@
     </remote_servers>
 
     <_functional_tests_helper_database_replicated_replace_args_macros>1</_functional_tests_helper_database_replicated_replace_args_macros>
+    <max_database_replicated_create_table_thread_pool_size>50</max_database_replicated_create_table_thread_pool_size>
```
Comment (tavplubix) [inline tests/config/config.d/database_replicated.xml:100]: Let's also enable it in the intagration test for Replicated db

**Review by tavplubix — APPROVED** [review]

**Pull request merged**